packages/cst-dts-gen/src/model.ts

```diff
@@ -0,0 +1,167 @@
+import {
+  Alternation,
+  Alternative,
+  GAstVisitor,
+  IProduction,
+  NonTerminal,
+  Option,
+  Repetition,
+  RepetitionMandatory,
+  RepetitionMandatoryWithSeparator,
+  RepetitionWithSeparator,
+  Rule,
+  Terminal,
+  TokenType
+} from "chevrotain"
+import {
+  assign,
+  flatten,
+  groupBy,
+  map,
+  mapValues,
+  some
+} from "@chevrotain/utils"
+
+export function buildModel(
+  productions: Record<string, Rule>
+): CstNodeTypeDefinition[] {
+  const generator = new CstNodeDefinitionGenerator()
+  return mapValues(productions, (rule) => generator.visitRule(rule))
+}
+
+export type CstNodeTypeDefinition = {
+  name: string
+  properties: PropertyTypeDefinition[]
+}
+
+export type PropertyTypeDefinition = {
+  name: string
+  type: PropertyArrayType
+  optional: boolean
+}
+
+export type PropertyArrayType = TokenArrayType | RuleArrayType
+export type TokenArrayType = { kind: "token" }
+export type RuleArrayType = {
+  kind: "rule"
+  name: string
+}
+
+class CstNodeDefinitionGenerator extends GAstVisitor {
+  visitRule(node: Rule): CstNodeTypeDefinition {
+    const rawElements = this.visitEach(node.definition)
+
+    const grouped = groupBy(rawElements, (el) => el.propertyName)
+    const properties = mapValues(grouped, (group, propertyName) => {
+      const allNullable = !some(group, (el) => !el.canBeNull)
+      return {
+        name: propertyName,
+        type: group[0].type,
+        optional: allNullable
```

Comment:
The logic for determining if a property is optional may be problematic.
For example: if error recovery is enabled, everything is theoretically optional.

Reply:
Also Chevrotain could probably be adjusted to return **partial** CST results
even without error recovery, so in such a CST, once again, everything is optional

Reply:
ah... _recovery_...  I hadn't considered that.

Perhaps the generator could check the `recoveryEnabled` flag on the parser instead of always marking them optional.
Or better yet we could change the visitor interface functions argument types to be `Partial<?>`.  That way the user can add the recovered node flag to the in parameter and use it as apart of a type guard to get the children type if no recovery took place.

i.e. something like this

```typescript
interface TestRuleCstNode extends CstNode {
  name: "testRule";
  children: TestRuleCstChildren;
}
  
type TestRuleCstChildren = {
  Token1: IToken[];
};

interface ICstNodeVisitor<IN, OUT> extends ICstVisitor<IN, OUT> {
  testRule(children: Partial<TestRuleCstChildren>, param?: IN): OUT;
}

type InputParam = { recoveredNode: boolean };

class CstVisitor extends parser.getBaseCstVisitorConstructor() implements ICstNodeVisitor<InputParam, unknown> {
  visit(cstNode: CstNode, params?: InputParam) {
    return super.visit(cstNode, { recoveredNode: !!cstNode.recoveredNode });
  }

  testRule(children: Partial<TestRuleCstChildren>, param?: InputParam): unknown {
    if (isNotRecovered(children, param!)) {
      return children.Token1; // <-- this is IToken[]
    } else {
      return children.Token1; // <-- this is IToken[] | undefined
    }
  }
}

function isNotRecovered<T>(children: Partial<T>, params: InputParam): children is T {
  return !params.recoveredNode;
}
```

Reply:
> Perhaps the generator could check the recoveryEnabled flag on the parser instead of always marking them optional.

There is a conceptual mismatch here, because the recoveryEnabled flag is a **runtime** artifact, while our generator 
happens at **design time**, basically a Chevrotain parser may accept a **runtime** argument for the `recoveryEnabled`.

> we could change the visitor interface functions argument types to be Partial<?>. That way the user can add the recovered node flag to the in parameter and use it as apart of a type guard to get the children type if no recovery took place.

I did not fully understand this workaround, however:
- I'd rather avoid complex logic in the CSTVisitor, particularly as I'd like to refactor the CSTVisitor at some point:
  - #1039 

Random Thoughts:

- With modern ECMAScript accessing "undefined" properties is not a big deal using [optional chaining](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Optional_chaining), so maybe its not a big deal if everything is optional?

- Can we expose a option for the user to decide if they want everything optional (e.g recovery scenario) or
  everything mandatory (compiler scenario)?

Reply:
>There is a conceptual mismatch here, because the recoveryEnabled flag is a runtime artifact, while our generator
happens at design time, basically a Chevrotain parser may accept a runtime argument for the recoveryEnabled.

The generator api takes in a parser instance currently so we could argue that we're generating types for that specific parser configuration.  But I guess we can expose an explicit option in the generator api if that is confusing.

>I'd rather avoid complex logic in the CSTVisitor

Fair enough.  Was thinking that having the compiler scenario types would be useful in a recovery scenario.  We can always revisit this later if there is plausible use case.